File: src/problems/Problem019.java
package problems;

public class Problem019 {

	private static int weekday = 1; // 0: Sun, 1: Mon, ... , 6: Sat
	private static int year = 1900;
	private static int month = 1;
	private static int day = 1;

	public static void main(String[] args) {
		int firstSundays = 0;
		while (year <= 2000) {
			if (year >= 1901 && day == 1 && weekday == 0)
				firstSundays++;
			progressDay();
		}
		System.out.println(firstSundays);
	}

	public static void progressDay() {
		weekday++;
		day++;
		if (weekday > 6) {
			weekday = 0;
		}
		if (day > daysInMonth(month, year)) {
			day = 1;
			progressMonth();
		}
	}

	public static void progressMonth() {
		month++;
		if (month > 12) {
			month = 1;
			year++;
		}
	}

	public static int daysInMonth(int month, int year) {
		switch(month) {
		case 9: // September
		case 4: // April
		case 6: // June
		case 11: // November
			return 30;
		case 2: // February
			return isLeapYear(year) ? 29 : 28;
		default:
			return 31;
		}
	}

	public static boolean isLeapYear(int year) {
		if (year % 4 == 0) {
			if (year % 100 == 0 && !(year % 400 ==0))
				return false;
			return true;
		} else
			return false;
	}

}

File: python/problem019.py

def main_OLD():
	firstSundays = 0
	weekday = 1
	month = 1
	year = 1900
	while year < 2001:	
		if weekday == 0 and year >= 1901:
			firstSundays += 1
		month += 1
		year += int(month / 12)
		month %= 12
		weekday += getDaysInMonth(month, year)
		weekday %= 7
	print(firstSundays)

def main():
	firstSundays = 0
	weekday = 1
	day = 1
	month = 1
	year = 1900
	while year < 2001:	
		if day == 1 and weekday == 0 and year >= 1901:
			firstSundays += 1
		day += 1
		weekday += 1
		weekday %= 7
		if day > getDaysInMonth(month, year):
			day = 1
			month += 1
			if month > 12:
				year += 1
				month = 1
	print(firstSundays)
	
def getDaysInMonth(month, year):
	if month in (9, 4, 6, 11):
		return 30
	elif month == 2:
		if year % 4 == 0:
			if year % 100 == 0 and not year % 400 == 0:
				return 28
			else:
				return 29
		else:
			return 28
	else:
		return 31

if __name__ == '__main__':
	main()